scripts/dev/upstream_merge/build_and_install_kernel.py

```diff
@@ -0,0 +1,291 @@
+    status_lines = [l for l in git_status_output.splitlines() if l.strip().startswith('U')]
+    status_lines = status_lines[:20]
+    status_block = "\n\nGit status (unmerged entries only):\n" + "\n".join(status_lines) if status_lines else ""
+    body = (
+        f"Merge conflict while merging {latest_tag} into {target_branch}\n"
+        f"Total conflicting files: {len(conflict_list)}\n"
+        "Conflicting files:\n" + "\n".join(shown_files) + truncated_note + status_block
+    )
+    return body
+
+# === Step 0: Check Required Packages ===
+def check_required_packages():
+    print("[INFO] Checking required Ubuntu packages...")
+    packages = config.required_packages
+    if not packages:
+        print("[INFO] No required packages specified in configuration.")
+        return
+    missing = []
+    for pkg in packages:
+        status, _ = run_command(f"dpkg -s {pkg}")
+        if status != 0:
+            missing.append(pkg)
+    if missing:
+        print(f"[INFO] Installing missing packages: {', '.join(missing)}")
+        status, output = run_command(f"sudo apt-get update && sudo apt-get install -y {' '.join(missing)}")
+        if status != 0:
+            raise RuntimeError(f"[ERROR] Package install failed: {output}")
+    else:
+        print("[INFO] All required packages are already installed.")
+
+# === Step 1: Detect or Build Toolchain ===
+# Uses shared toolchain detection utility from utils.toolchain module.
+
+# === Step 1.5: Ensure Kernel Source Repository ===
+# Clone kernel repository if missing or not a valid git directory.
+
+def ensure_kernel_source_repository():
+    print("[INFO] Ensuring kernel source repository exists...")
+    if not config.kernel_src_dir:
+        raise RuntimeError("[ERROR] No kernel_src_dir configured in kernel_build section")
+    if not config.repo_url:
+        raise RuntimeError("[ERROR] No repo_url configured for kernel repository")
+    # Convert to absolute path to prevent issues after directory changes
+    config.kernel_src_dir = os.path.abspath(config.kernel_src_dir)
+    kernel_src_dir = config.kernel_src_dir
+    parent_dir = os.path.dirname(kernel_src_dir)
+    if not os.path.exists(parent_dir):
+        print(f"[INFO] Creating parent directory: {parent_dir}")
+        os.makedirs(parent_dir, exist_ok=True)
+    if os.path.exists(kernel_src_dir):
+        if os.path.isdir(os.path.join(kernel_src_dir, '.git')):
+            print(f"[INFO] Kernel source repository present: {kernel_src_dir}")
+            return
+        print(f"[WARNING] {kernel_src_dir} exists but is not a git repository; replacing...")
+        shutil.rmtree(kernel_src_dir)
+    status, output = run_command(f"git clone {config.repo_url} {kernel_src_dir}")
+    if status != 0:
+        raise RuntimeError(f"[ERROR] Kernel repo clone failed: {output}")
+    print(f"[INFO] Cloned kernel repository to: {kernel_src_dir}")
+
+# === Step 2: Robust Upstream Merge with Conflict Handling ===
+def run_upstream_merge_script():
```

Comment:
Rather than call this `run_upstream_merge_script()`, you should name it something along the lines of `merge_upstream_kernel` or something similar.